components/renderers/controls/array/array-renderer.ts

```diff
@@ -2,210 +2,134 @@
+    protected static createControl(schemaPath: string, prop: string): IControlObject {
+        return {
+            "type": "Control",
+            "label": JSONForms.PathUtil.beautify(prop),
+            "scope": {
+                "$ref": schemaPath
             }
         }
-
-        return value;
     }
 
-    private mergeOptions(optionsA: any, optionsB: any): any {
-        return Object.keys(optionsB).reduce((mergedOpts, optionName) => {
-            if (mergedOpts.hasOwnProperty(optionName)) {
-                return mergedOpts
-            }
-            mergedOpts[optionName] = optionsB[optionName];
-            return mergedOpts;
-        }, optionsA);
+    protected createControlGroupPerItem(schemaPath: string, items: any, dataLength: number): IGroup[] {
+        return _.range(0, dataLength).map(index => {
+            let elements = _.keys(items['properties']).map(key =>
+                // path does not actually exists
+                ArrayRenderer.createControl(`${schemaPath}/items/${index}/properties/${key}`, key)
+            );
+            return ArrayRenderer.createGroup(elements);
+        });
     }
 
-    isApplicable(element: IUISchemaElement, subSchema: SchemaElement, schemaPath: string):boolean {
-        return element.type == 'Control' && subSchema !== undefined && subSchema.type == 'array';
+    protected createControlsForSubmit(items: any, schemaPath: string, submitElement: any, services: JSONForms.Services) {
+        let unboundControls = _.keys(items['properties']).map(prop =>
+            ArrayRenderer.createControl(`${schemaPath}/items/properties/${prop}`, prop)
+        );
+        let renderDescriptionsForSubmit = JSONForms.RenderDescriptionFactory.renderElements(unboundControls, this.renderService, services);
+        return renderDescriptionsForSubmit.map(renderDescription => {
+            renderDescription['instance'] = submitElement;
+            renderDescription['path'] = JSONForms.PathUtil.lastFragment(renderDescription['path']);
+            return renderDescription;
+        });
     }
 
-    render(element: IArrayControlObject, schema: SchemaElement, schemaPath: string, services: JSONForms.Services): JSONForms.IArrayControlRenderDescription {
+    protected generateControlDescriptions(items: any, schemaPath: string, currentDescriptions: JSONForms.IRenderDescription[], dataSize: number, services: JSONForms.Services) {
+        // TODO: this won't work for replace
+        // is there a better way to accomplish this?
+        if (currentDescriptions.length === dataSize) {
+            return currentDescriptions;
+        }
+        currentDescriptions.splice(0, currentDescriptions.length);
+        let controlGroups = this.createControlGroupPerItem(schemaPath, items, dataSize);
+        let renderDescriptions = JSONForms.RenderDescriptionFactory.renderElements(controlGroups, this.renderService, services);
+
+        return renderDescriptions.reduce((descriptions, renderDescription) => {
```

Comment:
wouldn't it be cleaner to use the descriptions variable instead of the currentdescriptions